File: src/exchange/binance.ts
/**
 * Binance 合约行情模块
 *
 * 方案：
 *   - WebSocket 订阅 bookTicker → 实时价格
 *   - REST API /fapi/v1/klines → 小时 K 线边界（小时开盘价）
 *   - 本地定时器检测小时切换
 */

import WebSocket from 'ws';
import { EventEmitter } from 'events';
import { config } from '../config.js';
import type { Candle } from '../types.js';

/** REST API 基地址 */
const REST_BASE = 'https://fapi.binance.com';

export class BinanceMarket extends EventEmitter {
  private ws: WebSocket | null = null;
  private _currentPrice = 0;
  private _lastHourCandle: Candle | null = null;
  private destroyed = false;

  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private reconnectAttempts = 0;
  private hourCheckTimer: ReturnType<typeof setInterval> | null = null;

  get currentPrice(): number {
    return this._currentPrice;
  }

  get lastHourCandle(): Candle | null {
    return this._lastHourCandle;
  }

  start(): void {
    if (this.destroyed) return;
    this.connectWs();
    this.startHourCheck();
  }

  // ---- WebSocket (bookTicker) ----

  private connectWs(): void {
    if (this.destroyed) return;
    this.cleanupWs();

    const url = `${config.binanceWsUrl}/ws`;
    console.log(`[WS] 连接中 (第${this.reconnectAttempts + 1}次)...`);

    const ws = new WebSocket(url);

    ws.on('error', (err: Error) => {
      console.error(`[WS] 错误: ${err.message}`);
    });

    ws.on('open', () => {
      console.log('[WS] 已连接');
      this.reconnectAttempts = 0;
      this.ws = ws;
      this.startPing();
      ws.send(JSON.stringify({
        method: 'SUBSCRIBE',
        params: [`${config.symbol}@bookTicker`],
        id: 1,
      }));
    });

    ws.on('message', (raw: WebSocket.RawData) => {
      try {
        const msg = JSON.parse(raw.toString());
        this.handleWsMessage(msg);
      } catch { /* ignore */ }
    });

    ws.on('close', (code: number) => {
      console.log(`[WS] 断开 (code=${code})`);
      this.cleanupWs();
      this.scheduleReconnect();
    });
  }

  private handleWsMessage(msg: Record<string, unknown>): void {
    // 跳过订阅确认消息
    if ('result' in msg) return;

    if (msg['e'] === 'bookTicker') {
      const bid = Number(msg['b'] ?? 0);
      const ask = Number(msg['a'] ?? 0);
      if (bid > 0 && ask > 0) {
        this._currentPrice = (bid + ask) / 2;
        this.emit('price', this._currentPrice);
      }
    }
  }

  // ---- REST API 获取小时 K 线 ----

  private async fetchHourCandle(): Promise<Candle | null> {
    try {
      const url = `${REST_BASE}/fapi/v1/klines?symbol=${config.symbol.toUpperCase()}&interval=1h&limit=1`;
      const res = await fetch(url);
      if (!res.ok) return null;
      const data = await res.json() as unknown[][];
      if (!data || data.length === 0) return null;
      const k = data[0];
      return {
        openTime: Number(k[0]),
        open: Number(k[1]),
        high: Number(k[2]),
        low: Number(k[3]),
        close: Number(k[4]),
      };
    } catch {
      return null;
    }
  }

  // ---- 小时边界检测（每秒检查一次） ----

  private startHourCheck(): void {
    this.stopHourCheck();
    // 启动时立即获取一次当前小时 K 线
    setTimeout(() => this.checkHourBoundary(), 1000);
    // 每秒检查
    this.hourCheckTimer = setInterval(() => this.checkHourBoundary(), 1000);
  }

  private stopHourCheck(): void {
    if (this.hourCheckTimer) {
      clearInterval(this.hourCheckTimer);
      this.hourCheckTimer = null;
    }
  }

  private async checkHourBoundary(): Promise<void> {
    const now = Date.now();
    const currentHourStart = Math.floor(now / 3600000) * 3600000;

    // 已经处理过这个小时的 K 线，跳过
    if (this._lastHourCandle?.openTime === currentHourStart) return;

    // 获取当前小时 K 线
    const candle = await this.fetchHourCandle();
    if (!candle) return;

    const prev = this._lastHourCandle;
    if (!prev || candle.openTime > prev.openTime) {
      const ts = new Date(candle.openTime).toLocaleTimeString('zh-CN', { hour12: false });
      console.log(`[K线] 新小时: ${ts}, 开仓价=${candle.open}`);
      this._lastHourCandle = candle;
      this.emit('hourCandle', candle);
    }
  }

  // ---- 心跳 / 重连 / 清理 ----

  private startPing(): void {
    this.stopPing();
    this.pingTimer = setInterval(() => {
      if (this.ws?.readyState === WebSocket.OPEN) {
        this.ws.ping();
      }
    }, 3 * 60 * 1000);
  }

  private stopPing(): void {
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
  }

  private scheduleReconnect(): void {
    if (this.destroyed) return;
    this.reconnectAttempts++;
    const delay = Math.min(5000 * Math.pow(2, this.reconnectAttempts - 1), 60_000);
    console.log(`[WS] ${Math.round(delay / 1000)} 秒后重连...`);
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this.connectWs();
    }, delay);
  }

  private cleanupWs(): void {
    this.stopPing();
    if (this.ws) {
      this.ws.removeAllListeners();
      try { this.ws.terminate(); } catch { /* ignore */ }
      this.ws = null;
    }
  }

  destroy(): void {
    this.destroyed = true;
    this.stopHourCheck();
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    this.cleanupWs();
  }
}

File: internal/market/binance.go
package market

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

const restBase = "https://fapi.binance.com"

type BinanceMarket struct {
	symbol string
	wsURL  string

	priceMu     sync.RWMutex
	currentPrice float64

	priceCh  chan PriceUpdate
	candleCh chan Candle

	ctx    context.Context
	cancel context.CancelFunc
	wg     sync.WaitGroup
}

func NewBinance(symbol, wsURL string) *BinanceMarket {
	return &BinanceMarket{
		symbol:   symbol,
		wsURL:    wsURL,
		priceCh:  make(chan PriceUpdate, 64),
		candleCh: make(chan Candle, 8),
	}
}

func (m *BinanceMarket) CurrentPrice() float64 {
	m.priceMu.RLock()
	defer m.priceMu.RUnlock()
	return m.currentPrice
}

func (m *BinanceMarket) PriceChan() <-chan PriceUpdate {
	return m.priceCh
}

func (m *BinanceMarket) CandleChan() <-chan Candle {
	return m.candleCh
}

func (m *BinanceMarket) Start(ctx context.Context) error {
	m.ctx, m.cancel = context.WithCancel(ctx)
	m.wg.Add(2)
	go m.runWS()
	go m.pollCandles()
	return nil
}

func (m *BinanceMarket) Stop() {
	if m.cancel != nil {
		m.cancel()
	}
	m.wg.Wait()
}

func (m *BinanceMarket) runWS() {
	defer m.wg.Done()
	for attempt := 0; ; attempt++ {
		if m.ctx.Err() != nil {
			return
		}
		m.connectWS(attempt)
		if m.ctx.Err() != nil {
			return
		}
		delay := time.Duration(math.Min(float64(5000*int64(1<<attempt)), 60_000)) * time.Millisecond
		select {
		case <-m.ctx.Done():
			return
		case <-time.After(delay):
		}
	}
}

func (m *BinanceMarket) connectWS(attempt int) {
	url := fmt.Sprintf("%s/ws", m.wsURL)
	dialer := websocket.Dialer{HandshakeTimeout: 10 * time.Second}
	conn, _, err := dialer.DialContext(m.ctx, url, nil)
	if err != nil {
		return
	}
	defer conn.Close()

	// Subscribe to bookTicker
	sub := map[string]interface{}{
		"method": "SUBSCRIBE",
		"params": []string{fmt.Sprintf("%s@bookTicker", m.symbol)},
		"id":     1,
	}
	if err := conn.WriteJSON(sub); err != nil {
		return
	}

	// Ping timer
	pingTicker := time.NewTicker(3 * time.Minute)
	defer pingTicker.Stop()

	go func() {
		for {
			select {
			case <-m.ctx.Done():
				return
			case <-pingTicker.C:
				conn.WriteMessage(websocket.PingMessage, nil)
			}
		}
	}()

	for {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			return
		}
		var msg map[string]interface{}
		if err := json.Unmarshal(raw, &msg); err != nil {
			continue
		}
		if _, ok := msg["result"]; ok {
			continue // subscription confirmation
		}
		if e, ok := msg["e"]; ok && e == "bookTicker" {
			bid := parseFloat(msg["b"])
			ask := parseFloat(msg["a"])
			if bid > 0 && ask > 0 {
				price := (bid + ask) / 2
				m.priceMu.Lock()
				m.currentPrice = price
				m.priceMu.Unlock()
				select {
				case m.priceCh <- PriceUpdate{Price: price, Time: time.Now().UnixMilli()}:
				default:
				}
			}
		}
	}
}

func (m *BinanceMarket) pollCandles() {
	defer m.wg.Done()

	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	var lastCandle *Candle

	for {
		select {
		case <-m.ctx.Done():
			return
		case <-ticker.C:
			candle, err := m.fetchLatestCandle()
			if err != nil || candle == nil {
				continue
			}
			if lastCandle == nil || candle.OpenTime > lastCandle.OpenTime {
				lastCandle = candle
				select {
				case m.candleCh <- *candle:
				default:
				}
			}
		}
	}
}

func (m *BinanceMarket) fetchLatestCandle() (*Candle, error) {
	url := fmt.Sprintf("%s/fapi/v1/klines?symbol=%s&interval=1h&limit=1", restBase, m.symbol)
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("status %d", resp.StatusCode)
	}
	var data [][]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}
	if len(data) == 0 {
		return nil, nil
	}
	k := data[0]
	return &Candle{
		OpenTime: int64(parseFloat(k[0])),
			Open:     parseFloat(k[1]),
			High:     parseFloat(k[2]),
			Low:      parseFloat(k[3]),
			Close:    parseFloat(k[4]),
	}, nil
}

func parseFloat(v interface{}) float64 {
	switch val := v.(type) {
	case float64:
		return val
	case string:
		var f float64
		fmt.Sscanf(val, "%f", &f)
		return f
	}
	return 0
}

// FetchCandles fetches historical klines for backtesting
func FetchCandles(symbol, interval string, days int) ([]Candle, error) {
	cpd := candlesPerDay(interval)
	limit := min(days*cpd, 1500)
	url := fmt.Sprintf("%s/fapi/v1/klines?symbol=%s&interval=%s&limit=%d",
		restBase, symbol, interval, limit)

	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("binance API error: %d", resp.StatusCode)
	}

	var raw [][]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&raw); err != nil {
		return nil, err
	}

	candles := make([]Candle, len(raw))
	for i, k := range raw {
		candles[i] = Candle{
			OpenTime: int64(parseFloat(k[0])),
			Open:     parseFloat(k[1]),
			High:     parseFloat(k[2]),
			Low:      parseFloat(k[3]),
			Close:    parseFloat(k[4]),
		}
	}

	needed := days * cpd
	for len(candles) < needed && len(candles) >= 1500 {
		earliest := candles[0].OpenTime
		url2 := fmt.Sprintf("%s/fapi/v1/klines?symbol=%s&interval=%s&limit=1500&endTime=%d",
			restBase, symbol, interval, earliest-1)
		resp2, err := http.Get(url2)
		if err != nil {
			break
		}
		var raw2 [][]interface{}
		if err := json.NewDecoder(resp2.Body).Decode(&raw2); err != nil {
			resp2.Body.Close()
			break
		}
		resp2.Body.Close()
		if len(raw2) == 0 {
			break
		}
		more := make([]Candle, len(raw2))
		for i, k := range raw2 {
			more[i] = Candle{
				OpenTime: int64(parseFloat(k[0])),
				Open:     parseFloat(k[1]),
				High:     parseFloat(k[2]),
				Low:      parseFloat(k[3]),
				Close:    parseFloat(k[4]),
			}
		}
		candles = append(more, candles...)
		if len(more) < 1500 {
			break
		}
	}

	if len(candles) > needed {
		candles = candles[len(candles)-needed:]
	}
	return candles, nil
}

func candlesPerDay(interval string) int {
	m := map[string]int{
		"1m": 1440, "3m": 480, "5m": 288, "15m": 96, "30m": 48,
		"1h": 24, "2h": 12, "4h": 6, "6h": 4, "8h": 3, "12h": 2, "1d": 1,
	}
	if v, ok := m[interval]; ok {
		return v
	}
	return 24
}
